packages/gatsby-source-contentful/src/normalize.js

```diff
@@ -49,41 +49,53 @@ const fixId = id => {
 }
 exports.fixId = fixId
 
+const shouldBeWalked = (object, alreadyWalkedObjectRefs) =>
+  object && typeof object === `object` && !alreadyWalkedObjectRefs.has(object)
+
 const fixIds = object => _fixIds(object)
 
-// Recursively walk the object model and find any property named `sys`. If it
+// Walk the object model and find any property named `sys`. If it
 // contains an `id` then make sure the id is a string and if it starts with a
 // number, prefix it with (an arbitrarily chosen) `c`, for "contentful".
 // The `front` tracks which objects have been visited to prevent infinite
 // recursion on cyclic structures.
-const _fixIds = (object, front = new Set()) => {
-  if (!object || typeof object !== `object`) {
-    return
-  }
+const _fixIds = object => {
+  const objectsToProcess = [object]
+  const alreadyWalkedObjectRefs = new Set(objectsToProcess)
 
-  if (Array.isArray(object)) {
-    object.forEach(v => _fixIds(v, front))
-    return
-  }
+  while (objectsToProcess.length !== 0) {
+    const current = objectsToProcess.pop()
 
-  if (front.has(object)) {
-    return
-  }
+    if (!current || typeof current !== `object`) {
+      continue
+    }
```

Comment:
You can move these three lines out of this loop and apply them only to the initial `object`. All other values will go through `shouldBeWalked` before queued so this check is redundant for all but the first value.